File: app/src/main/java/com/coinninja/coinkeeper/service/client/NetworkingApiClient.java
package com.coinninja.coinkeeper.service.client;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.IOException;

import androidx.annotation.NonNull;
import okhttp3.Protocol;
import okhttp3.Request;
import retrofit2.Call;
import retrofit2.Response;

public abstract class NetworkingApiClient {
    public static final int ERROR_CODE = 418;

    protected Response executeCall(Call call) {
        Response response = null;
        try {
            response = call.execute();
        } catch (IOException e) {
            e.printStackTrace();
            response = createTeaPotErrorFor(call, e.getMessage());
        }

        return response;
    }

    public Response createTeaPotErrorFor(Call call, String message) {
        return Response.error(okhttp3.ResponseBody.create(null, ""), new okhttp3.Response.Builder()
                .code(NetworkingApiClient.ERROR_CODE)
                .message(message)
                .protocol(Protocol.HTTP_1_1)
                .request(new Request.Builder().url(call.request().url()).build())
                .build());

    }

    @NonNull
    protected JsonObject createQuery(String property, String[] data) {
        Gson gson = new Gson();
        JsonElement element = gson.toJsonTree(data);
        return createQuery(property, element);
    }

    @NonNull
    protected JsonObject createQuery(String property, JsonElement element) {
        JsonObject json = new JsonObject();
        json.add(property, element);

        JsonObject terms = new JsonObject();
        terms.add("terms", json);

        JsonObject query = new JsonObject();
        query.add("query", terms);
        return query;
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/service/client/NetworkingApiClient.kt
package com.coinninja.coinkeeper.service.client

import app.dropbit.annotations.Mockable
import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import okhttp3.Protocol
import okhttp3.Request
import retrofit2.Call
import retrofit2.Response
import java.io.IOException

@Mockable
abstract class NetworkingApiClient {


    protected fun createQuery(property: String, data: Array<String>): JsonObject {
        val gson = Gson()
        val element = gson.toJsonTree(data)
        return createQuery(property, element)
    }

    protected fun createQuery(property: String, element: JsonElement): JsonObject {
        val json = JsonObject()
        json.add(property, element)

        val terms = JsonObject()
        terms.add("terms", json)

        val query = JsonObject()
        query.add("query", terms)
        return query
    }

    protected fun <T> executeCall(call: Call<T>): Response<T> {
        val response: Response<T>

        response = try {
            call.execute()
        } catch (e: IOException) {
            e.printStackTrace()
            createTeaPotErrorFor(call, e.message)
        }

        return response
    }

    protected fun <T> createTeaPotErrorFor(call: Call<T>, message: String?): Response<T> {
        return Response.error(okhttp3.ResponseBody.create(null, ""), okhttp3.Response.Builder()
                .code(ERROR_CODE)
                .message(message ?: "")
                .protocol(Protocol.HTTP_1_1)
                .request(Request.Builder().url(call.request().url()).build())
                .build())
    }

    companion object {
        const val ERROR_CODE = 418
    }
}
